File: practica_control_2.py
import numpy as np
import matplotlib.pyplot as plt
import control as ctrl
from control import forced_response  # Importar la función similar a lsim

# Parámetros del sistema
num = [3]
den = [1, 2, 3]
theta = 2       # Retardo puro
ts = 0.1        # Tiempo de muestreo

# Sistema sin retardo
Gsr = ctrl.tf(num, den)

# Aproximación del retardo
num_pade, den_pade = ctrl.pade(theta, 1)
delay_sys = ctrl.tf(num_pade, den_pade)

# Sistema total con retardo
Gsr_with_delay = ctrl.series(delay_sys, Gsr)

# Vector de tiempo
t = np.arange(0, 30 + ts, ts)
N = len(t)

# Construcción de la señal arbitraria
sizev = int(N * 0.2)
s1 = np.ones(sizev) * 2
s2 = np.zeros(sizev)
s3 = np.linspace(5, 2, sizev)
s4 = np.linspace(3, 7, sizev)
s5 = np.ones(sizev + 1) * 7
arbsig = np.concatenate([s1, s2, s3, s4, s5])

# Asegurar que la señal tenga la misma longitud que el tiempo
if len(arbsig) < len(t):
    arbsig = np.append(arbsig, [arbsig[-1]] * (len(t) - len(arbsig)))
elif len(arbsig) > len(t):
    arbsig = arbsig[:len(t)]

# Simulación de la respuesta del sistema
t_out, y_out = forced_response(Gsr_with_delay, T=t, U=arbsig)

# Gráficas
plt.figure(figsize=(10, 6))

plt.subplot(2, 1, 1)
plt.plot(t, arbsig)
plt.title("Señal arbitraria")
plt.xlabel("Tiempo [s]")
plt.ylabel("Amplitud")
plt.grid(True)
plt.legend(["Señal en el tiempo"], loc="best")

plt.subplot(2, 1, 2)
plt.plot(t_out, y_out)
plt.plot(t, arbsig)
plt.title("Respuesta a la señal")
plt.xlabel("Tiempo [s]")
plt.ylabel("Amplitud")
plt.grid(True)
plt.legend(["Respuesta del sistema", "Arbitraria"], loc="best")
plt.tight_layout()
plt.show()

File: practica_control_2.m
clc; clear;

num = 3;
den = [1 2 3];
ts = 0.1;
theta = 2;
Gs = tf(num,den); %T. continuo
Gsr = tf(num,den,'InputDelay', theta);
[impulso, t1] = impulse(Gs);
[escalon,t2] = step(Gs);
[impulso_ret,t3] = impulse(Gsr);
[escalon_ret,t4] = step(Gsr);
max_imp = max(impulso);
max_esc = max(escalon);
max_imp_ret = max(impulso_ret);
max_esc_ret = max(escalon_ret);

[~, id_imp] = min(abs(impulso - max_imp));
t_aprox_imp = t1(id_imp);
[~, id_esc] = min(abs(escalon - max_esc));
t_aprox_esc = t2(id_esc);
[~, id_imp_ret] = min(abs(impulso_ret - max_imp_ret));
t_aprox_imp_ret = t3(id_imp_ret);
[~, id_esc_ret] = min(abs(escalon_ret - max_esc_ret));
t_aprox_esc_ret = t4(id_esc_ret);

% subplot(2,1,1);
% plot(t1,impulso);
% hold on;
% scatter(t_aprox_imp,max_imp);
% subplot(2,1,2);
% plot(t2, escalon);
% hold on;
% scatter(t_aprox_esc,max_esc);

% subplot(2,1,1);
% plot(t3,impulso_ret);
% hold on;
% scatter(t_aprox_imp_ret,max_imp_ret);
% subplot(2,1,2);
% plot(t4, escalon_ret);
% hold on;
% scatter(t_aprox_esc_ret,max_esc_ret);

t = 0:ts:30;
t = t';
N = length(t);

sizev = int32(N*0.2);

%senal arbitraria
s1 = ones(1, sizev)*2;
s2 = zeros(1,sizev);
s3 = linspace(5,2,sizev);
s4 = linspace(3,7,sizev);
s5 = ones(1,sizev + 1)*7;


arbsig = [s1,s2,s3,s4,s5]';
u2 = [t arbsig];

subplot(2,1,1);
plot(t,arbsig);
grid on;
title("Señal arbitraria");
xlabel("Tiempo");
ylabel("Amplitud");
legend("Señal en el tiempo", 'Location', 'best');
subplot(2,1,2);
lsim(Gsr,arbsig,t);
grid on;
title("Respuesta a la señal");
xlabel("Tiempo");
ylabel("Amplitud");
legend("Respuesta del sistema", 'Location', 'best');
